lib/request.js

```diff
@@ -242,9 +243,12 @@ module.exports = {
    */
 
   get hostname() {
-    const host = this.host;
-    if (!host) return '';
-    return host.split(':')[0];
+    if (!this.parsedUrl) {
+      const host = this.host;
+      if (!host) return '';
+      this.parsedUrl = new URL(`${this.protocol}://${this.host}`);
```

Comment:
Use WHATWG parser for hostname when `host` start with `[`, otherwise keep the old way to get high performance.